asserts/account_key_test.go

```diff
@@ -419,3 +421,61 @@ func (aks *accountKeySuite) TestPrerequisites(c *C) {
 		PrimaryKey: []string{"acc-id1"},
 	})
 }
+
+func (aks *accountKeySuite) TestKeyProofHappy(c *C) {
+	kproof, err := asserts.SignWithoutAuthority(asserts.KeyProofType,
+		map[string]interface{}{
+			"nonce": "NONCE",
+			"key":   aks.pubKeyEncoded,
+		}, nil, aks.privKey)
+	c.Assert(err, IsNil)
+
+	// roundtrip
+	a, err := asserts.Decode(asserts.Encode(kproof))
+	c.Assert(err, IsNil)
+
+	kproof2, ok := a.(*asserts.KeyProof)
+	c.Assert(ok, Equals, true)
+
+	// standalone signature check
+	err = asserts.SignatureCheck(kproof2, aks.privKey.PublicKey())
+	c.Check(err, IsNil)
+
+	c.Check(kproof2.Nonce(), Equals, "NONCE")
+}
+
+func (aks *accountKeySuite) TestKeyProofDecodeInvalid(c *C) {
+	encoded := "type: key-proof\n" +
+		"nonce: NONCE\n" +
+		"key:\n    KEY\n" +
+		"body-length: 0\n" +
+		"sign-key-sha3-384: " + aks.privKey.PublicKey().ID() + "\n\n" +
+		"AXNpZw=="
+
+	invalidTests := []struct{ original, invalid, expectedErr string }{
+		{"nonce: NONCE\n", "", `"nonce" header is mandatory`},
+		{"nonce: NONCE\n", "nonce: \n", `"nonce" header should not be empty`},
+		{"key:\n    KEY\n", "", `"key" header is mandatory`},
+		{"key:\n    KEY\n", "key: \n", `"key" header should not be empty`},
+		{"key:\n    KEY\n", "key: $$$\n", `cannot decode public key: .*`},
+	}
+
+	for _, test := range invalidTests {
+		invalid := strings.Replace(encoded, test.original, test.invalid, 1)
+		invalid = strings.Replace(invalid, "KEY", strings.Replace(aks.pubKeyEncoded, "\n", "\n    ", -1), 1)
+		_, err := asserts.Decode([]byte(invalid))
+		c.Check(err, ErrorMatches, keyProofErrPrefix+test.expectedErr)
+	}
+}
+
+func (aks *accountKeySuite) TestKeyProofDecodeKeyIDMismatch(c *C) {
+	invalid := "type: key-proof\n" +
```

Comment:
OK.  So, the point of this assertion is as follows.  account-key assertions now (as of a couple of weeks ago) only have public-key-sha3-384 as their primary key, and not also account-id: this means that we should put some effort into ensuring that people can't register account-keys to their account that they don't actually own, because otherwise they would be able to block the legitimate owner from registering it.  (Yes, in practice at the moment external considerations will make that somewhat unlikely, because the key will probably only be generated right before registering it and will be in a directory that only the generating user can read, but in principle it is a public key and I strongly believe that we should not rely on such non-cryptographic external factors.)

Of course the key-proof assertion itself only proves that whoever signed the assertion has access to the private key.  That's its purpose, and that's why this assertion is only intended for use on the wire and not to be stored long-term.  (In fact, since it contains a nonce, it really could not legitimately be stored.)  In that respect it's like the existing assertions I mentioned which are used in device authentication.

As mentioned above, when you ask the store to register an account-key (that is, to create an account-key assertion for the given public key and account-id), part of that exchange involves the client proving that it possessing the key.  This was alluded to in the Snapcraft UX changes prepared in Heidelberg, although not in much detail.  A pretty standard approach to proving that you possess a key is to sign a message containing a nonce and the public key data.  My first pass at this was to encode those two items in a JSON object.  However, this was a bit ad-hoc, and it wasn't going to work very well with Samuele's in-progress signing tool, which can only construct and sign assertions.

Putting this all together, my intent is that registering an account-key will work as follows (outlined [here](https://docs.google.com/document/d/1DOgDCT5VK9n7PsQToNPCOaS3wGVZVqZDurq2iIUxw7M/edit#heading=h.gqh6exn7uj62) although I haven't filled in the assertion details yet):

 * Retrieve a nonce from the store.

 * Construct an assertion like this and sign it with the same key:

        type: key-proof
        nonce: <nonce>
        key: <encoded key data>

 * Send a macaroon-authorised request to SCA/dev/api/account/account-key whose body is `{"key_data": "<encoded key data>", "proof": "<key-proof assertion>"}`.

 * If the nonce was previously unused, both assertions are valid and correctly signed, and the key in the key-proof assertion matches the requested key_data, then the store creates the account-key assertion.

So the key-proof assertion is only used in transit, and goes with the account-key creation request (or other similar things we might need in future, so Samuele suggested giving it a rather generic name).

Does that clarify matters?